spec.bs

```diff
@@ -461,6 +473,146 @@ An <dfn>internal observer</dfn> is a [=struct=] with the following [=struct/item
   [[#promise-returning-operators]] that make use of this, for example.</p>
 </div>
 
+<div algorithm>
+  To <dfn for=Observable>convert to an Observable</dfn> given an {{any}} |value|, run these steps:
+
+  Note: We split this algorithm out from the Web IDL {{Observable/from()}} method, so that
+  spec prose can <a for=Observable lt="convert to an observable">convert</a> an {{Observable}}
+  without going through the Web IDL bindings.
+
+  Note: The resolution of value to its descrete types happens before
+  [=Observable/subscribe callback=] is called. This means mutations of values, such as adding
+  the iterable protocols to the object, will not take effect between the creation of the returned
+  observable, and when it is subscribed to.
+
+    1. If [$Type$](|value|) is not [=Object=], [=exception/throw=] a {{TypeError}}.
+
+       Note: This prevents primitive types from being coerced into iterables (e.g., String).
+
+       Issue: See if this is even the behavior we want. See <a
+       href=https://github.com/WICG/observable/issues/125>WICG/observable#125</a>
+
+    1. <i id=from-observable-conversion><b>From Observable</b></i>: If |value|'s [=specific type=]
+       is an {{Observable}}, then return |value|.
+
+    1. <i id=from-async-iterable-conversion><b>From async iterator</b></i>: Let
+       |asyncIteratorMethodRecord| be [=?=] [$GetMethod$](|value|, {{%Symbol.asyncIterator%}}).
+
+       Note: We have to use [$GetMethod$] directly instead of [$GetIterator$], because
+       [$GetIterator$] makes it impossible to distinguish between (a) the case where |value| simply
+       doesn't implement any iterable protocol, and (b) the case where |value| implements one of the
+       iterator protocols, and its method getter throws an exception.
+
+       Note: This step is web-observable, and re-throws any errors that the {{%Symbol.iterator%}}
+       method *getter* might have thrown.
+
+    1. If |asyncIteratorMethodRecord|'s \[[Value]] is undefined or null, then jump to the step
+       labeled <a href=#from-iterable-conversion>From iterable</a>.
+
+    1. If [$IsCallable$](|asyncIteratorMethodRecord|'s \[[Value]]) is false, then
+       [=exception/throw=] a {{TypeError}}.
+
+       Otherwise, let |nextAlgorithm| be the following steps, given |subscriber| and |iterator|:
+
+         1. If |iterator|'s \[[Done]] is true, then: run |subscriber|'s {{Subscriber/complete()}}
+            method and abort these steps.
+
+         1. Let |nextRecord| be [$IteratorStepValue$](|iterator|).
```

Comment:
First off, we might want to use `IteratorNext()` instead of `IteratorStepValue()`. I'll investigate this. Second, we need to handle throw completions here, since we're not doing that currently. Probably we want to queue a microtask to call `error()` on abrupt completions, with the completion value.

Reply:
Done in https://github.com/WICG/observable/pull/160/commits/4a34553cc13fe1e7b06c33db5b02e8051dca71bf. I've left `IteratorStepValue()` alone here since that seems fine, but I've handled throw completions and the Done value _after_ we pull the next value, since `[[Done]]` should always be false the very first time, so we don't need to check it before we step forward.

Reply:
Actually I think we can't use `IteratorStepValue()`, since it expects the `next()` method to return an `IteratorResult` object with a `value` property. But for async iterators, per https://tc39.es/ecma262/#table-async-iterator-required, the `next()` method is expected to return a `Promise` (and if it isn't a `Promise` we wrap it in anyways), which resolves _to_ an `IteratorResult`. So I think we do have to use `IteratorNext()` like the Streams Standard does, and react to the returned Promise, and inside the reaction branch can we grab the value from the object etc.

Reply:
Done in https://github.com/WICG/observable/pull/160/commits/1b92dc5c14c6d80adb6a55e16231c1072ec7393d.